.github/workflows/build.yml

```diff
@@ -41,7 +41,7 @@ jobs:
           - os: ubuntu-22.04
             pg_major: 15
             config: Release
-            test_tfm: netcoreapp3.1
+            test_tfm: net6.0
```

Comment:
Wasn't netcoreapp3.1 here just so we test against netstandard?

Reply:
I don't think so - for that netcoreapp2.0 would be better... I think it's just a leftover from the netcoreapp3.1 removal, for me the point here is more to test our lowest supported non-netfx TFM...

After all, we no longer support netstandard2.0 on anything other than netfx (since all non-netfx TFMs under net6.0 are now out of support), and even netfx isn't supported in the sense that we run tests on it...

Note that targeting netcoreapp3.1 started causing [issues](https://github.com/npgsql/npgsql/actions/runs/5495588536/jobs/10015074743?pr=5149) in the actual build, which is why I started looking at this in the first place...

Reply:
> Wasn't netcoreapp3.1 here just so we test against netstandard?

IIRC this was indeed the case.

Just to make sure: Do we currently run tests that consume our .NET Standard 2.0 library and run tests against it?

Reply:
> Do we currently run tests that consume our .NET Standard 2.0 library and run tests against it?

After this PR no, we don't. But the point is that there's no more support for netstandard2.0 with any TFM except for netfx, which we don't test (because there were issues etc.)...

Reply:
Wouldn't the idea be that because netcoreapp isn't one of our supported tfms it gets the ns2.0 binary. Testing at least our ns2.0 ifdefs? (obviously the platform still isn't netfx)

Reply:
This would effectively mean that our netfx support is "tested" by running the tests with netcoreapp3.1 - not sure how much real coverage that adds.

Since it seems you guys think it's important, I tried bringing back testing on netcoreapp3.1 - but there's a problem. Since there's no more netstandard2.1, the tests project now uses netstandard2.0; since that contains our own polyfill versions of certain attributes (e.g. [AllowNull], [DoesNotReturn]), usage of these attributes in the test code are now ambiguous: Npgsql's polyfills are exposed because we have InternalsVisibleTo to the test project, but also the netcoreapp3.1 BCL has them. We can use [extern alias](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/extern-alias) to give Npgsql an alias and disambiguate, but at that point we have to prefix all Npgsql namespaces with that alias in all the tests files, which is a bit much. Or, of course, we could try removing the InternalsVisible too (which is also a bit much). Or we could keep targeting netstandard2.1, but doing that just to support testing sucks.

If anyone has some bright idea about how to work around this, let me know... Otherwise I don't think it's too bad to say "we don't disallow netfx but it's not fully tested" (again, it wasn't fully tested anyway).

Reply:
Ahah yeah there is this workaround by yours truly ;) https://github.com/dotnet/sdk/issues/30151

> I have another shitty solution, segregate all the polyfills to another project which is only IVT to the library and not the tests...

Reply:
OK, came up with another idea - the problematic attributes are now defined in the namespace Npgsql.Polyfills, and we have a global usings for the for netstandard2.0 only in the csproj. This way, the tests don't use that namespace and never see the internal attributes.

Lots of work/complexity for what it is :/

Reply:
The problem is that... there are test failures. I don't think this indicates bugs, but I don't have time right now to go through it. Does anyone else want to take a look?

Reply:
> The problem is that... there are test failures. I don't think this indicates bugs, but I don't have time right now to go through it. Does anyone else want to take a look?

That's not bugs but a feature 🐝

.NET Framework doesn't support cancellation of `SslStream.ReadAsync`, which is why we immediately break the connection for .NET Standard 2.0.

https://github.com/npgsql/npgsql/blob/c36dc1f1f56ad8aa8e05e3297ab02b2417cd6196/src/Npgsql/Internal/NpgsqlReadBuffer.cs#L216-L222

Since .NET Core 3.1 support .NET Standard 2.1, until now we were using it instead of 2.0 version, so our tests weren't affected by this. And with 2.1 being out of the picture...

Reply:
Anything before .NET 6 is effectively out of support. I suggest we don't make more work for ourselves here.

We can always reconsider on a case-by-case basis if we receive a number of complaints. Though it would always just be a courtesy given the support policy that modern .NET has.

Reply:
@Brar I do understand the concern - if we want to be conservative here, we can just continue targeting netstandard2.1 as before; though note that dropping net5.0 and netcoreapp3.1 was the same kind of decision: people still on those TFMs may start getting behavior changes if they upgrade to Npgsql 8 (we just don't have tests showing them at the moment).

I'm still leaning to saying "we support only supported TFMs", which effectively means we guarantee nothing for them.

One interesting idea may be to drop netstandard2.0, but to target net48 (or whatever). That would mean you can no longer use unsupported TFMs at all, so rather than getting runtime behavior changes you can't upgrade; on the other hand, I'm guessing most people wouldn't run into the behavior changes even when they suddenly get our netstandard2.0 assembly instead of the netstandard2.1 one.

So I think our options are:
* Keep netstandard2.1, although effectively all the covered TFMs are unsupported.
* Drop netstandard2.0 and target net48
* Ignore the failing tests for the netcoreapp3.1 build of the tests.

Reply:
I don't think we can drop ns2.0 because of mono and unity no?

Reply:
I cannot say that I really care which path we choose.
The less overall work our choice causes for us, the better.

Reply:
> I don't think we can drop ns2.0 because of mono and unity no?

Don't these support net48 (or whatever netfx TFM)?

Reply:
I mean this crap https://learn.microsoft.com/en-us/nuget/reference/target-frameworks. I have no idea how it exactly works but just like unity they support netstandard but not always specifically netfx.

<img width="738" alt="Screenshot 2023-07-10 at 18 49 14" src="https://github.com/npgsql/npgsql/assets/4218809/ea253b27-ad02-4b36-92be-ea6e199dcf8a">

Reply:
OK, I guess that option's out then...

Reply:
If you want more bedtime reading I can highly recommend reading these to fall asleep quicky:

https://github.com/dotnet/designs/blob/main/accepted/2020/net5/net5.md
https://github.com/dotnet/designs/blob/main/accepted/2021/net6.0-tfms/net6.0-tfms.md

Reply:
> [ADDED] Stop running the tests with netcoreapp3.1

After this, I have to say that I do care.

I cannot see a point in supporting something we consider unsupported (to the point that we ourselves don't know which feature actually works and which does not). That would be "Drop netstandard2.0" for me.

Reply:
> I cannot see a point in supporting something we consider unsupported (to the point that we ourselves don't know which feature actually works and which does not).

That has already been the case for several release with .NET Framework; it's been "supported" in that we target netstandard2.0, but we weren't running tests on it. Using netcoreapp3.1 as a "proxy" for netfx isn't really testing.

But I think this "partial support " mode for .NET Framework is OK. We're don't fully support it, but we don't actively block people from using it; it mostly works (nobody has been complaining!) but we don't commit to fixing issues which are found on .NET Framework only. IIRC it has been this way for several releases, and I think it has been OK...

Reply:
Yeah, maybe.

FWIW I'd probably go with "Ignore the failing tests for the netcoreapp3.1 build of the tests". That way we'd at least have some implicit documentation on what works and what doesn't.

Reply:
This is such an awesome PR, "I'll just go ahead and remove netstandard2.1, it'll be a small easy thing" :rofl::rofl::rofl:

Reply:
Let's push this pr a bit.

To summarize:
1. Contrary to our expectations we didn't run tests on netstandard2.0 but instead on netstandard2.1
2. We still would like to run tests on netstandard2.0 in some capacity

With that in mind I see 3 options:
1. Keep running tests on netcoreapp3.1 as is (far from ideal since it's going to constantly fail and be noisy in general)
2. Rewrite/disable some tests around cancellation for netcoreapp3.1 (probably going to require quite a bit of time + any new test we add around cancellation will have to be disabled)
3. Change the check in `NpgsqlReadBuffer` so we fail if and only if we're on netstandard2.0 and not on .NET Core (the only option to do this I found for now is by parsing `RuntimeInformation.FrameworkDescription`)

Reply:
See https://github.com/npgsql/npgsql/pull/5149#issuecomment-1685129639 as well (let's revisit this in few weeks)